### DISCUSSION TITLE
Verwendung von ioBroker zum Zusammenfassen von Verbrauchsdaten

### DISCUSSION THREAD

**Discussion by ledermann** [solectrus/solectrus#3782] [Show and tell]
Derzeitig läuft der [Beta-Test für die Visualisierung selbstdefinierte Verbraucher](https://github.com/orgs/solectrus/discussions/3749) in SOLECTRUS. Die Idee ist, den Verbrauch ausgewählter Geräte mit separaten Stromzählern (z.B. von Shelly) zu messen und in SOLECTRUS darzustellen.

Eine Möglichkeit, die Verbrauchdaten zu SOLECTRUS zu bekommen, ist die Verwendung des [Shelly-Collectors](https://github.com/solectrus/shelly-collector). Dieser fragt die Messwerte der Shellys permanent ab und speichert sie in der InfluxDB, die von SOLECTRUS gelesen wird.

Das funktioniert gut, hat aber zwei Nachteile, wenn man viele Geräte einbinden möchten:

- Für jeden Shelly muss ein separater Shelly-Collector als Docker-Container laufen, die Konfiguration wird schnell unübersichtlich
- Jeder einzelne Shelly taucht separat in SOLECTRUS auf, was bei Kleinverbrauchern wenig hilfreich ist, außerdem stößt man schnell an das [Limit von 10 Verbrauchern](https://github.com/orgs/solectrus/discussions/3749#discussioncomment-11557463).

## Gruppenbildung als Lösung

Eine Lösung kann sein, die Verbrauchswerte der Shellys thematisch in Gruppen zusammenzufassen und nur die Gruppen-Summen an SOLECTRUS zu senden.

Als Gruppen könnte man z.B. "Küche", "IT", "Heizen", "Wäsche" oder was auch immer definieren. Ich möchte kurz zeigen, wie man das mit ioBroker machen kann. Es gibt sicherlich viele andere Möglichkeiten, das zu realisieren. Man muss auch nicht zwingend ioBroker verwenden, mit HomeAssistant oder ganz anderen Tools geht das sicherlich auch.

Grob skizziert bin ich so vorgegangen:

1. Jeder Shelly meldet seine Messwerte eigenständig per MQTT an ioBroker
2. In ioBroker wird ein Skript ausgeführt, das die Messwerte der einzelnen Shellys zusammenfasst und die Summe als neues Objekt speichert
3. Die Summe wird dann mit dem [MQTT-Collector](https://github.com/solectrus/mqtt-collector) abgeholt und in die InfluxDB von SOLECTRUS geschrieben, den Shelly-Collector benötigt man dann gar nicht mehr.

## Schritt 1: Shelly konfigurieren

In jedem Shelly wird MQTT aktiviert und konfiguriert, sodass er seine Messwerte an ioBroker sendet. Hier ein Beispiel:

![shelly-mqtt](https://github.com/user-attachments/assets/eb0236f7-748f-4495-a673-2e1f32571801)

Der Shelly sendet dann bei jeder Änderung des Verbrauches eine Nachricht an ioBroker. Dieser speichert die Nachricht als Objekt, beispielsweise im Pfad `mqtt/0/Shelly/iMacPro'.

Das macht man für jeden einzelnen Shelly. Ob es klappt, sollte man im ioBroker überprüfen, auch das dortige Log ist hilfreich, um Fehler zu finden.

## Schritt 2: ioBroker-Skript erstellen

In ioBroker installiert man zunächst die [Script-Engine](https://github.com/ioBroker/ioBroker.javascript). Diese ermöglicht es, eigene Skripte zu erstellen, die auf Ereignisse reagieren und Aktionen ausführen.

Ich habe folgendes Script geschrieben, das die Verbrauchswerte von sieben Shellys in zwei Gruppen zusammengefasst (Küche und IT) und die Summe berechnet. Das Ergebnis wird jeweils in einem separaten Objekt gespeichert und alle 5 Sekunden aktualisiert.

Das Script ist **nicht** geeignet, um es einfach so zu kopieren, da es spezifisch für meine Konfiguration ist. Aber es zeigt, wie das prinzipiell funktioniert. 

```javascript
// Definition der Kategorien und zugehörigen Geräte
const shellyCategories = {
    Kitchen: {
        devices: [
            'mqtt.0.Shelly.Fridge.status.switch:0',
            'mqtt.0.Shelly.Washer.status.switch:0',
            'mqtt.0.Shelly.Dishwasher.status.switch:0'
        ],
        targetId: 'javascript.0.shellyKitchen.totalPower'
    },
    IT: {
        devices: [
            'mqtt.0.Shelly.Mac.status.switch:0',
            'mqtt.0.Shelly.Display.status.switch:0',
            'mqtt.0.Shelly.Synology.status.switch:0',
            'mqtt.0.Shelly.KabelFritz.status.switch:0',
            'mqtt.0.Shelly.iMacPro.status.switch:0'
        ],
        targetId: 'javascript.0.shellyIT.totalPower'
    }
};

// Ziel-Datenpunkte erstellen
Object.values(shellyCategories).forEach(({ targetId }) => {
    if (!existsState(targetId)) {
        createState(targetId, 0, { name: `Total Power`, type: 'number', role: 'value', unit: 'W' });
    }
});

// Funktion zur Extraktion und Berechnung der Summe
function parsePower(state) {
    try {
        return state ? JSON.parse(state)?.apower || 0 : 0;
    } catch (e) {
        log(`Error parsing state: ${state} - ${e}`, 'error');
        return 0;
    }
}

// Funktion zur Berechnung der Leistung für eine Kategorie
function updateCategoryPower(categoryName) {
    const category = shellyCategories[categoryName];
    let total = 0;

    category.devices.forEach((id) => {
        const state = getState(id)?.val;
        total += parsePower(state);
    });

    setState(category.targetId, total, true);
}

Object.entries(shellyCategories).forEach(([categoryName]) => {
    // Initiale Berechnung
    updateCategoryPower(categoryName);

    // Periodische Aktualisierung
    setInterval(() => updateCategoryPower(categoryName), 5000); // 5 Sekunden
});
```

Den MQTT-Broker weist man nun an, dass er die Datenpunkte `javascript.0.*` bereitstellt. Dadurch kommt man von außen an die berechneten Werte heran.

## Schritt 3: MQTT-Collector konfigurieren

Den Shelly-Collector benötigt man gar nicht mehr. Stattdessen richtet man den [MQTT-Collector](https://github.com/solectrus/mqtt-collector) so ein, dass er die berechneten Summen abholt:

```env
MAPPING_0_TOPIC=javascript/0/shellyKitchen/totalPower
MAPPING_0_MEASUREMENT=Kitchen
MAPPING_0_FIELD=power
MAPPING_0_TYPE=integer

MAPPING_1_TOPIC=javascript/0/shellyIT/totalPower
MAPPING_1_MEASUREMENT=IT
MAPPING_1_FIELD=power
MAPPING_1_TYPE=integer
```

**Comment by bgriese0** [comment]
Hallo zusammen,

ich finde die Lösung individuelle Verbraucher darzustellen super. Ich versuche gerade über den MQTT Collector Werte aus ioBroker in Solectrus zu bekommen, scheitere aber daran, dass sie im Dashboard angezeigt werden. 

Die Verbraucher in Influx sind definiert:

`      - INFLUX_SENSOR_CUSTOM_POWER_01`

Das Mapping sollte ebenfalls richtig sein:

```
- MAPPING_12_TOPIC=0_userdata/0/SmartHome/Leistung_USV
- MAPPING_12_MEASUREMENT=it
- MAPPING_12_FIELD=power
- MAPPING_12_TYPE=float
```

Die Werte kommen im Collector auch an:

```
  topic = 0_userdata/0/SmartHome/Leistung_USV
  message = 97.5
  => it:power = 97.5
```

Leider wird mir nach einem neustart und auch nach Zurücksetzen der Tageswerte kein Reiter mit Verbrauchern im Dashboard gezeigt. Woran kann es liegen?

Viele Grüße
Björn

**Reply by ledermann** [reply]
Die Darstellung der Selbstdefinierten Verbrauchern befindet sich momentan noch im Betatest, ist also nicht in der stabilen Version (derzeitig v0.18.3) enthalten. Du kannst das Feature aber nutzen, wenn du auf das Docker-Tag `pr-3747` umstellst. Mehr dazu hier:
https://github.com/orgs/solectrus/discussions/3749
